### PR TITLE
VYZN: Matrix Messaging Integration

### PR DESCRIPTION
**Pull request by clecherbauer**
This pull request implements the lib "matrix-widget-api" which allows us to send control-messages to an embedded (iframe) bldrs instance.

Messages bldrs can receive:
- ai.bldrs-share.LoadModel (bldrs loads a ifc-model from github)
- ai.bldrs-share.SelectElements (bldrs select elements from the current model in the viewer)

Messages bldrs will emit:
- ai.bldrs-share.ElementsSelected (triggers if elements are added to the selection)
- ai.bldrs-share.ElementsDeSelected (triggers if elements are removed from the selection)

### CODE REVIEW

**Comment by oo-bldrs** [comment]
The title of this PR should be changed to reflect the contents, which if I'm guessing correctly is: Matrix Messaging Integration?

**Review by oo-bldrs — CHANGES_REQUESTED** [review]
I've added some comments regarding what's present so far.

File: src/WidgetApi/ApiConnection.js
Comment (clecherbauer) [reply src/WidgetApi/ApiConnection.js:3]: good point

File: src/index.jsx
Comment (clecherbauer) [reply src/index.jsx:9]: The Problem here is that we need to initialize this class at a place which doesn't get reinitialized. I debugged this and the outcome was that any component gets reloaded at routing time. Since both ends of matrix-widget-api relies on a specific order of events there is no stable communication when initialized in routed components.

Comment (clecherbauer) [reply src/index.jsx:9]: but i'm not specialized in building react-apps maybe you guys know a better way of doing this :)

Comment (pablo-mayrgundter) [reply src/index.jsx:9]: I don't know the answer, but I think this is worth solving.  Happy to help debug together.  Have you tried in src/Share.jsx or its parent components?  It only rerenders on new models, but I think that's probably desirable here.

File: src/WidgetApi/WidgetApi.js
Comment (pablo-mayrgundter) [inline src/WidgetApi/WidgetApi.js:16]: Can this actually throw?

File: src/WidgetApi/WidgetApi.js
Comment (pablo-mayrgundter) [inline src/WidgetApi/WidgetApi.js:8]: const

File: src/store/useStore.js
Comment (pablo-mayrgundter) [inline src/store/useStore.js:9]: I suspect this will grow quickly.  Please make a slice for it called ApiSlice with field URI

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]
Thanks, nice start!

Let's leave this in draft until we get tests and styles.  Left you some initial commements.  I'll help with the global init when we can chat.

**Comment by pablo-mayrgundter** [comment]
Here's our style guide:

  https://github.com/bldrs-ai/Share/wiki/Dev:-Style

Please use our eslint config or run ```yarn lint``` manually

File: src/WidgetApi/WidgetApi.js
Comment (clecherbauer) [reply src/WidgetApi/WidgetApi.js:16]: This is the source of this function: [https://stackoverflow.com/questions/326069/how-to-identify-if-a-webpage-is-being-loaded-inside-an-iframe-or-directly-into-t](https://stackoverflow.com/questions/326069/how-to-identify-if-a-webpage-is-being-loaded-inside-an-iframe-or-directly-into-t)

File: src/index.jsx
Comment (clecherbauer) [reply src/index.jsx:9]: i tested it again today and it only works stable at this place

**Comment by pablo-mayrgundter** [comment]
Per discussion, I'd try useMemo or useCallback for the init logic.  Assuming that works, the right place is probably in src/Share.jsx.

**Comment by pablo-mayrgundter** [comment]
Hey Christian, how's this going?

**Comment by clecherbauer** [comment]
@pablo-mayrgundter i still need to write tests. Could you do another review now before is start with that?

File: src/WidgetApi/WidgetApi.js
Comment (pablo-mayrgundter) [inline src/WidgetApi/WidgetApi.js:8]: export default class WidgetApi

File: src/WidgetApi/WidgetApi.js
Comment (pablo-mayrgundter) [inline src/WidgetApi/WidgetApi.js:10]: pls add type annotations

File: src/WidgetApi/ApiEventsRegistry.js
Comment (pablo-mayrgundter) [inline src/WidgetApi/ApiEventsRegistry.js:92]: This seems brittle.. maybe array->set for each and compare sets?

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]
Basically looking good.

Would you please update the description of the PR to summarize the changes and a link to the staging instance you're testing with.

**Review by pablo-mayrgundter — APPROVED** [review]
Hey Christian, tests look good!

It's been a long time since you started, so there's a chance this may conflict with some in-flight changes.  Looks like you need to merge.  If that goes fine, I'll merge.

Thanks!  Happy Holidays!
Pablo

**Comment by clecherbauer** [comment]
hey @pablo-mayrgundter, i did the merge

**Comment by pablo-mayrgundter** [comment]
Hmm, I'm not seeing the netlify dialog change.  can you try updating smth like that yarn.lock and see if that triggers it?

**Comment by clecherbauer** [comment]
OK, i did update yarn.lock. But does not seem like it triggered the deployment. We really need to have the current version of this code deployed since the viewer in our production is loaded from this preview-instance.

**Comment by oo-bldrs** [comment]
I manually flushed the cache and retried the build. Does the deploy preview reflect your changes?

https://deploy-preview-455--bldrs-share.netlify.app/

**Comment by clecherbauer** [comment]
yes, thank you

**Comment by Adrian62D** [comment]
Hi Pablo, going directly for #525 would be easier. However, if you wish you can first merge this #455. But we should not invest more time #455.

**Comment by pablo-mayrgundter** [comment]
Per discussion with Adrian yesterday, Christian, please do finish this out.  I think we're good to go but pending a successful merge and final review (hopefully a no-op).

**Comment by clecherbauer** [comment]
@pablo-mayrgundter Hi Pablo, @aozien has rebased and fixed my changes. I think we're ready for your final review now :)

**Comment by pablo-mayrgundter** [comment]
Looks like we got broken somehow :(

I tried loading a model and got this error.  I see some maybe related changes in the last update.

image

**Comment by aozien** [comment]
@pablo-mayrgundter  It turns out the problem was caused by the (yarn.lock) file re-generated in commit (a8ed57d), since that did some minor-version updates to some packages, which caused some functionality to break.
The yarn lock file is now changed to match exactly the one on share/main for the already existing packages.

**Comment by pablo-mayrgundter** [comment]
Works for me now, thanks.

Found another issue.. permalinks for camera and cutplane are broken :(

https://deploy-preview-455--bldrs-share.netlify.app/share/v/p/index.ifc/89/112/139/154/621#c:43.83,54.18,47.56,-47,15,-5.73::p:x=0

When I follow this link, everything after the index.ifc is removed after load.  Same on a loaded model.

Please fix that and add an integration test for it.

Thank you!

**Comment by aozien** [comment]
@pablo-mayrgundter  this should be fixed now as well :)

**Comment by pablo-mayrgundter** [comment]
Finally!  Congrats guys!

**Pull request merged**
